### DISCUSSION TITLE
Automated removal of molecules with 'valency issues' in virtual screening pipeline

### DISCUSSION THREAD

**Discussion by lynden95** [rdkit/rdkit#6174] [Q&A]
Hey RDKit gang,

I am trying to take a list of smiles, standardise them, convert them to 2D then to 3D, merge into an SDF file and run molecular docking as one automated pipeline. However, there are often molecules that are in smiles lists that, during conversion to 2D-3D, have issues either embedding or with valencies. I am trying to find a way of removing these molecules and capturing the information of those molecules I have discarded (or the ones that are left!) so I can, at a later stage, append a docking score to a smiles list for ML.

I can successfully remove molecules that don't embed with the following code:

`mol = Chem.MolFromSmiles(ST_SMILES)
            mol = Chem.AddHs(mol)
            ps = AllChem.ETKDGv2()
            ps.useRandomCoords = True
            result = AllChem.EmbedMolecule(mol, ps)
            if type(result) is int:
                if result == -1:
                    print(ID)
                    continue`

This prints the ID of the problematic molecule, which can be saved to a list, although I haven't quite worked this out yet (working on it!).

The issue is that compounds will successfully embed, and be converted to PDB and SDF format, even if there are valency issues. It would be great to identify these compounds earlier, fix these issues, or even remove them from the original .csv file containing the smiles string, as early as possible. The following code incorporates 'good' molecules with correct valency into a merged_sdf file for docking, but I lose all information about which molecules are lost. This messes up the numbering, which is important as I need to append a docking score at a later date!

`mols = []
    for s in out_sdfs:
        suppl = Chem.SDMolSupplier(s)
        for  i, mol in enumerate(suppl):
            if mol is None:
                continue
        else:
            mols += [mol]
        os.remove(s)
    with Chem.SDWriter(merged_sdf) as w:
        for a, mol in enumerate(mols):
            if mol is None:
                continue
            else:
                w.write(mol)`

I think one of 2 things needs to happen - the problematic molecules are removed long before this stage, or information on which compounds are being omitted from the final merged_sdf list needs to be stored somewhere. Any help is much appreciated!

**Comment by DavidACosgrove** [comment]
It would be a lot easier to see what's going on if you enclose your code in three backticks so the formatting is maintained.
From what I can make out, you should probably open the supplier in non-sanitize mode, sanitize as a separate step inside the loop and catch any exceptions from the sanitisation to report the error.

**Comment by bp-kelley** [comment]
By default, MolFromSmiles returns None on failure so you can do something similar to the mol supplier

```
>>> m = Chem.MolFromSmiles("not a smiles")
[14:04:53] SMILES Parse Error: syntax error while parsing: not
[14:04:53] SMILES Parse Error: Failed parsing SMILES 'not' for input: 'not'
>>> print(m)
None
>>> 
```

```
for ST_SMILES, ID in ...:
    mol = Chem.MolFromSmiles(ST_SMILES) 
    if mol is None:
        print(ID)
        continue
    mol = Chem.AddHs(mol) 
    ps = AllChem.ETKDGv2() 
    ps.useRandomCoords = True 
    result = AllChem.EmbedMolecule(mol, ps) 
    if result == -1: 
        print(ID) 
        continue
   ...
```
